### PR TITLE
fix(idpe-17943): spans in futures stream need parent

### PR DESCRIPTION
**Pull request by wiedld**
Closes https://github.com/influxdata/idpe/issues/17943

~Not entirely sure why this fixes it. The passed span (into the future stream) is later pinned (pin-projected) and used to create many span recorders. These new recorders fail to find the parent_span (to attach properly) -- unless we make this change here. This change is to have a parent span recorder (not just a parent span) in the lifetime/scope where the futures stream is created.~ 

Believe we need a span recorder at each level, in order to properly attached to the parent (both the start time and the export to the collector).

--- 

See attached to parent:

Screen Shot 2023-07-28 at 2 02 07 PM

For reference, here is BEFORE this change:
(unattached to any parent, and with clock warning)

Screen Shot 2023-07-28 at 2 15 28 PM

### CODE REVIEW

**Review by domodwyer — COMMENTED** [review]
Can we add a test that covers what this fixes? I can imagine those clones getting "cleaned up" fairly easily!

The "why" is nothing to do with lifetimes (there are none in this!) - creating the `SpanRecorder` _starts_ the span:

https://github.com/influxdata/influxdb_iox/blob/main/trace/src/span.rs#L171-L173

So this incantation causes the span to be _started_ which in turn means it will be emitted (and therefore child spans are linkable to it)

File: ingester/src/server/grpc/query.rs [hunk 1]
```diff
@@ -375,6 +375,11 @@ fn encode_response(
 
         let mut batch_iter = partition.into_record_batches().into_iter().peekable();
 
+        // TEST: demonstrate that this is not sufficient to include parent span
+        if let Some(span) = span.as_mut() {
+            span.start = Some(chrono::Utc::now());
```
Comment (wiedld) [inline ingester/src/server/grpc/query.rs:380]: This is insufficient to attach to the collector. See both the test case, and then here:

Screen Shot 2023-07-31 at 12 11 09 PM

Comment (wiedld) [inline ingester/src/server/grpc/query.rs:380]: This is insufficient to attach to the collector. See both the test case, and then here:

Screen Shot 2023-07-31 at 12 11 09 PM

Comment (wiedld) [reply ingester/src/server/grpc/query.rs]: Next step, will revert this change. Then demonstrate that the original/proposed change will pass the test case. 😄

File: ingester/src/server/grpc/query.rs [hunk 2]
```diff
@@ -351,6 +351,8 @@ fn encode_response(
     span: Option<Span>,
     frame_encoding_duration_metric: Arc<DurationHistogram>,
 ) -> impl Stream<Item = Result<FlightData, FlightError>> {
+    let span = SpanRecorder::new(span.clone()).span().cloned();
```
[the remark this answers is not in the document]
Comment (wiedld) [reply ingester/src/server/grpc/query.rs:354]: The SpanRecorder both sets the start time, and does the export to the collector on drop.

File: ingester/src/server/grpc/query.rs (hunk 1 above)
Comment (domodwyer) [reply ingester/src/server/grpc/query.rs]: To be clear - I never said "we shouldn't use the `SpanRecorder`" - I was just explaining why it had nothing to do with lifetimes :) 

It's definitely preferable to use the `SpanRecorder` which encapsulates the timing / clock source, than to hand wrangle it like in this commit :+1:

**Review by domodwyer — APPROVED** [review]
this frame tracing has been such an ordeal! nice to get it going :rocket:

**Pull request merged**